### PR TITLE
Dockerize API

### PR DESCRIPTION
**Pull request by mmurto**
Split the work in #803 to separate PRs for API and UI. Draft waiting for #805 and #804 to be merged.

### CODE REVIEW

**Comment by mmurto** [comment]
Not ready for review but trying to figure out why checks don't run.

File: API.Dockerfile [hunk 1]
```diff
@@ -0,0 +1,23 @@
+# SPDX-FileCopyrightText: 2024 HH Partners
```
Comment (sschuberth) [inline API.Dockerfile:1]: Should that be Double Open by now?

File: API.Dockerfile [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+WORKDIR /app/dos
+
+COPY package.json ./package.json
+COPY apps/api/package.json ./apps/api/package.json
+COPY packages/database/package.json ./packages/database/package.json
+COPY packages/s3-helpers/package.json ./packages/s3-helpers/package.json
+COPY packages/validation-helpers/package.json ./packages/validation-helpers/package.json
+COPY package-lock.json ./package-lock.json
+
+RUN npm ci
```
Comment (sschuberth) [inline API.Dockerfile:17]: Just checking: No `pnpm` here and below?

**Review by sschuberth — COMMENTED** [review]
Just a bunch of remarks and questions, otherwise AFAICT looking good.

**Comment by sschuberth** [comment]
Ah, sorry, no idea why this popped up in my review notifications if it's not ready for review yet.

File: API.Dockerfile (hunk 2 above)
Comment (mmurto) [reply API.Dockerfile:17]: DOS doesn't use pnpm.

File: Worker.Dockerfile [hunk 3]
```diff
@@ -1,4 +1,4 @@
-# SPDX-FileCopyrightText: 2024 HH Partners
+# SPDX-FileCopyrightText: 2024 Double Open Oy
```
Comment (sschuberth) [inline Worker.Dockerfile:1]: The changes in this file are unrelated to the commit topic and should be split out to a separate commit.

File: API.Dockerfile (hunk 1 above)
Comment (sschuberth) [reply API.Dockerfile:1]: Ping.

File: .dockerignore [hunk 4]
```diff
@@ -4,7 +4,8 @@
 
 Dockerfile
 .dockerignore
-node_modules/
+**/node_modules/
+.git
```
Comment (sschuberth) [inline .dockerignore:8]: I guess this should have a trailing slash to make the entry only apply to directories.

File: Worker.Dockerfile (hunk 3 above)
Comment (mmurto) [reply Worker.Dockerfile:1]: So do you think copyright change should always be a separate commit, and not done when the file is modified?

File: .dockerignore (hunk 4 above)
Comment (mmurto) [reply .dockerignore:8]: .dockerignore differs from .gitignore in that `.git` only targets the root level, it's not recursive without the wildcard, so it will always target only the git directory in the root.

File: Worker.Dockerfile (hunk 3 above)
Comment (sschuberth) [reply Worker.Dockerfile:1]: Uhm, not necessarily, although it's probably a good practice. But it seemed to me the file is not modified (besides adding a trailing newline) in that commit. Let me re-check...

Comment (mmurto) [reply Worker.Dockerfile:1]: Is separated it.

File: .dockerignore (hunk 4 above)
Comment (mmurto) [reply .dockerignore:8]: In theory it would match a file `.git` in root, but as we maintain the project with Git, it will always have a `.git` directory in root, and most operating systems don't allow files and directories to have the same path, so in practice this will never match a file with a path of `.git`.

File: Worker.Dockerfile [hunk 5]
```diff
@@ -38,4 +38,4 @@ RUN npm run build:scanner_worker
 # Ensure scancode is accessible from the PATH
 ENV PATH="/venv/bin:${PATH}"
 
-CMD cd apps/scanner_worker && npm run start
\ No newline at end of file
+CMD cd apps/scanner_worker && npm run start
```
Comment (sschuberth) [inline Worker.Dockerfile:41]: Still, there is no change in this file except for adding the newline, so it shouldn't be part of this commit.

File: .dockerignore (hunk 4 above)
Comment (sschuberth) [reply .dockerignore:8]: I agree. Would you mind to still make it read `.git/` for consistency with all other entries which also have a trailing slash if a directory is supposed to be matched?

File: Worker.Dockerfile (hunk 3 above)
Comment (sschuberth) [reply Worker.Dockerfile:1]: This is the commit that should have the added newline in the end.

**Review by sschuberth — CHANGES_REQUESTED** [review]

**Review by sschuberth — APPROVED** [review]
